.github/workflows/workflow.yml

```diff
@@ -1,183 +1,334 @@
+
+      - name: Use zstd for faster cache restore
+        if: ${{ matrix.os == 'windows-latest' }}
+        shell: cmd
+        run: echo C:\Program Files\Git\usr\bin>>"%GITHUB_PATH%"
+
+      - name: Cache deps
+        uses: actions/cache@v3
+        with:
+          path: venv
+          key: deps-${{ matrix.os }}-${{ matrix.numpy }}-${{ matrix.python }}-${{ hashFiles('setup.py') }}
 
       - name: Cache build
-        id: restore-build
-        uses: actions/cache@v2
+        uses: actions/cache@v3
         with:
-          path: ${{ env.pythonLocation }}
-          key: build-${{ env.pythonLocation }}-${{ hashFiles('setup.py') }}-${{ github.sha }}
+          path: venv/**/[oO]pen[fF]isca*
+          key: build-${{ matrix.os }}-${{ matrix.numpy }}-${{ matrix.python }}-${{ hashFiles('setup.py') }}-${{ github.sha }}
 
-      - name: Check NumPy typing against latest 3 minor versions
-        run: for i in {1..3}; do VERSION=$(${GITHUB_WORKSPACE}/.github/get-numpy-version.py prev) && pip install numpy==$VERSION && make check-types; done
+      - name: Run Extension Template tests
+        run: |
+          ${{ matrix.activate }}
+          make test-extension
 
   lint-files:
-    runs-on: ubuntu-20.04
-    needs: [ build ]
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-20.04, windows-latest]
+        numpy: [1.20.3]
+        python: [3.8.10, 3.7.9]
+        include:
+        - os: ubuntu-20.04
+          activate: source venv/bin/activate
+        - os: windows-latest
+          activate: .\venv\Scripts\activate
+    runs-on: ${{ matrix.os }}
+    name: lint-files-${{ matrix.os }}-${{ matrix.numpy }}-${{ matrix.python }}
+    needs: [ deps ]
     env:
       TERM: xterm-256color
 
     steps:
-      - uses: actions/checkout@v2
+      - uses: actions/checkout@v3
         with:
           fetch-depth: 0 # Fetch all the tags
 
       - name: Set up Python
         uses: actions/setup-python@v4
         with:
-          python-version: 3.7.12
+          python-version: ${{ matrix.python }}
 
-      - name: Cache build
-        id: restore-build
-        uses: actions/cache@v2
+      - name: Use zstd for faster cache restore
+        if: ${{ matrix.os == 'windows-latest' }}
+        shell: cmd
+        run: echo C:\Program Files\Git\usr\bin>>"%GITHUB_PATH%"
+
+      - name: Cache deps
+        uses: actions/cache@v3
         with:
-          path: ${{ env.pythonLocation }}
-          key: build-${{ env.pythonLocation }}-${{ hashFiles('setup.py') }}-${{ github.sha }}
+          path: venv
+          key: deps-${{ matrix.os }}-${{ matrix.numpy }}-${{ matrix.python }}-${{ hashFiles('setup.py') }}
 
       - name: Run linters
-        run: make lint
+        run: |
+          ${{ matrix.activate }}
+          make clean lint
 
   check-version:
-    runs-on: ubuntu-20.04
-    needs: [ test-core, test-country-template, test-extension-template, check-numpy, lint-files ] # Last job to run
```

Comment:
The idea behind these dependencies is we want to give feedback to contributors on the version number only after they have passed all tests, so they don't have to do it twice after changes happened to the main branch during the time they took to fix the tests 🙂 Perhaps this rationale could be added as a comment!